typed-racket-lib/typed-racket/base-env/prims-contract.rkt

```diff
@@ -250,9 +251,17 @@
 ;;  make-predicate
 ;;  cast
 
-;; Helper to construct syntax for contract definitions
+;; Helpers to construct syntax for contract definitions
+;; make-contract-def-rhs : Type-Stx Boolean Boolean -> Syntax
 (define (make-contract-def-rhs type flat? maker?)
-  (contract-def-property #'#f `#s(contract-def ,type ,flat? ,maker? untyped)))
+  (define contract-def `#s(contract-def ,type ,flat? ,maker? untyped))
+  (contract-def-property #'#f (λ () contract-def)))
```

Comment:
I just thought of something else. If it were a promise, than an "accidental" too-early force would cache the wrong result, and the contract-generation pass would generate a wrapping `Any` contract even though by that time the `tc/#%expression` has put the correct type in the table. Using lambda means it might recompute a few more things, but that gets rid of the danger that a too-early application will ruin the result for future function applications of the lambda.